### ISSUE TITLE
Syntax error on hello_world.txp, line 1

### ISSUE THREAD

**Issue by ghuysmans** [patoline/patoline#16]
Hello,

After I manually `make install`'ed _earley_, I could finally build patoline. However, I can't get it to build the simple _hello_world_ example:
```
~/prog/patoline/examples$ patoline hello_world.txp 
[PPP] pa_patoline --main --build-dir .patobuild hello_world.txp > .patobuild/hello_world.ml
Parse error: file "hello_world.txp", line 1, character 0.
[RMV] rm -f .patobuild/hello_world.ml
PPP failed on ".patobuild/hello_world.ml"...
```

I'm running OCaml 4.05.0 (the version suggested by the README), let me know if you need more information.

Thanks!

**Comment by rlepigre** [comment]
Hi! Thanks for trying!

That's really weird, I can't reproduce. Are you running Linux? Have you installed the required libraries by hand? With Opam?

**Comment by ghuysmans** [comment]
I'm running Debian. I've installed the dependencies by running `opam install dune earley.2.0.0 camlzip sqlite3 imagelib` but it complained about `earley.2.0.0` (it could only install up to version 1.0.2, maybe I should install opam 2...), so I cloned it manually then ran `make install` in its directory before building patoline (successfully, except that I can't get it to parse anything, even a single comment).

**Comment by rlepigre** [comment]
OK, I see. I can't see any problem with that (but yeah, you should install opam 2 since the central repository has moved to opam 2 format). Installing `earley` by hand should definitely not pose a problem since this is what I do on my machine.

Frankly, I don't know what to tell you!

Could you try to run `pa_patoline --ascii file.txp` on simple examples, for example an empty file, or just some text? This command is simply a preprocessor converting Patoline syntax into OCaml.

And for good measures, can you check that `make tests` does not fail in the `earley` repository?

**Comment by ghuysmans** [comment]
I guess that all tests succeed:
```
~/prog/earley$ make tests
~/prog/earley$ echo $?
0
```

... but `pa_patoline` fails:
```
~/prog/patoline/examples$ pa_patoline --ascii hello_world.txp 
Parse error: file "hello_world.txp", line 1, character 0.
~/prog/patoline/examples$ touch empty.txp
~/prog/patoline/examples$ pa_patoline --ascii empty.txp 
Parse error: file "empty.txp", line 1, character 0.
```

I'll try this on another machine before trying to update to opam 2...

**Comment by rlepigre** [comment]
OK, thanks. It is possible that something really weird happens, for example that our database for unicode properties has not been installed in the right place. We rely on opam for that, so it may well be related to the fact that you don't use opam 2.

Anyway, thank you for testing things out!

**Comment by ghuysmans** [comment]
I found this, I guess that's what you call the Unicode database:
```
~/prog/patoline/examples$ strace pa_patoline --ascii empty.txp |& grep open\\\|stat
(...)
stat("./DefaultGrammar.tgy", 0x7ffeb50df790) = -1 ENOENT (No such file or directory)
stat(".patobuild/DefaultGrammar.tgy", 0x7ffeb50df790) = -1 ENOENT (No such file or directory)
stat("/usr/local/share/patoline/grammars/DefaultGrammar.tgy", 0x7ffeb50df790) = -1 ENOENT (No such file or directory)
```

Maybe the parser could issue a warning instead of rejecting any input (even a single newline after a comment) when no grammar can be loaded.

Anyway, when I manually symlink the grammar into the current directory, the generated program raises an unhandled exception about some missing `/usr/local/share/patoline/fonts` directory, then another one about a missing `hyphen` thingy in the same directory. Now I'm certain that my opam version is too old!

**Comment by rlepigre** [comment]
Cool, thank you again!

I'll try to see what I can do to make installation more reliable. We recently switched to dune, which is the new cool way to build stuff in OCaml (I wonder how long it will last). But it has annoying limitations, especially when it comes to "configure" steps: there is none. So there is no way to now in advance where stuff will actually be installed...

Anyway, installation works okay if you have opam 2, but as you noticed, if you have an obsolete version of opam (or no opam at all) then you are screwed. And that should not be the case.

**Comment by ghuysmans** [comment]
For what I do, dune works great (and is blazing fast), but at first it wasn't compatible with some nice PPXes. I wonder if it's changed and I was surprised you could use it at all (since pa_xxx are are completely different beast).

About opam, it looks like some dependencies are missing: at least `decompress` and `earley`. I also had an error about `ocamlbuild`, it tried using the system compiler's package, which obviously didn't work. I guess their sandbox disallows access to packages you don't depend on.

**Comment by rlepigre** [comment]
Yeah, I guess `decompress` (like `early`) is now only available trough opam 2. However, we do not directly depend on `ocamlbuild`, so the error you had was probably related to one of our dependencies.

As far as I know, opam did not do proper sandboxing before version 2, but I may be mistaken.

**Comment by ghuysmans** [comment]
I probably misread the error message, sorry. Thanks for your help!

**Comment by jorsn** [comment]
Maybe it would be good to extract the prefix directory from the build at compile time instead of relying on runtime `opam var share` or hardcoding `/usr/local/` in `unicodelib/config.ml` and `patconfig/patDefault.ml`. Is this easily possible?

In my [nix expression](https://paste.sr.ht/~jorsn/0db971616b31488efe7e465be0d4d6b334a36174) I had the same problem and solved it by substituting `/usr/local` by the real install directory in these two files.

**Comment by rlepigre** [comment]
I'd love to be able to do that, but I don't think that kind of stuff is possible with dune. This is a real problem with dune (at least for me), I reported it here: https://github.com/ocaml/dune/issues/1534.

**Comment by jorsn** [comment]
Yes, it seems one cannot detect it reliably, only in some cases (see [my comment in ocaml/dune#1185](https://github.com/ocaml/dune/issues/1185#issuecomment-691508200_)).

One could use `opam config var prefix` (you require opam anyway) in a dune action or add a `PREFIX` environment variable to the GNUMakefile and use it in dune actions, or add a make variable and directly generate a configure file.

In the first case, the `--prefix` option to `dune install` would be disallowed, and in the other two, the primary build system would be make and not dune.

One could only rely on a set `PREFIX` variable in the configure script and fail if it is not set. If it is documented, people hopefully will set `PREFIX` and `--prefix` to the same value.
